File: src/main/python/nQueens/NQueens.py
class Solution:
    def solveNQueens(self, n: int) -> list[list[str]]:
        board = [["."] * n for _ in range(n)]
        result = []
        self.dfs(0, board, result)
        return result

    def dfs(self, row, board, result):
        if row == len(board):
            result.append(["".join(row) for row in board])
            return
        for col in range(len(board)):
            if self.is_safe(row, col, board):
                board[row][col] = "Q"
                self.dfs(row + 1, board, result)
                board[row][col] = "."

    def is_safe(self, row, col, board):
        n = len(board)
        for i in range(n):
            if board[i][col] == "Q":
                return False
            if row >= i and col + i < n and board[row - i][col + i] == "Q":
                return False
            if row >= i and col - i >= 0 and board[row - i][col - i] == "Q":
                return False
        return True

    def test(self):
        tests = [
            [4, [[".Q..", "...Q", "Q...", "..Q."], ["..Q.", "Q...", "...Q", ".Q.."]]],
            [1, [["Q"]]],
        ]
        for test in tests:
            n, expect_result = test
            result = self.solveNQueens(n)
            print(result)
            print(result == expect_result)


if __name__ == "__main__":
    Solution().test()

File: src/main/kotlin/me/bossm0n5t3r/leetcode/nQueens/NQueens.kt
package me.bossm0n5t3r.leetcode.nQueens

class NQueens {
    class Solution {
        fun solveNQueens(n: Int): List<List<String>> {
            val result = Result()
            val board = Array(n) { CharArray(n) { '.' } }
            dfs(n, board, 0, result)
            return result.result.map { it.toList() }
        }

        private data class Result(val result: MutableSet<List<String>> = mutableSetOf())

        private fun dfs(
            n: Int,
            board: Array<CharArray>,
            row: Int,
            result: Result,
        ) {
            if (row == n) {
                result.result.add(board.map { it.joinToString("") })
                return
            }
            (0 until n).forEach { c ->
                if (isSafe(n, board, row, c)) {
                    board[row][c] = 'Q'
                    dfs(n, board, row + 1, result)
                    board[row][c] = '.'
                }
            }
        }

        private fun isSafe(n: Int, board: Array<CharArray>, row: Int, col: Int): Boolean {
            (0 until n).forEach { i ->
                if (board[i][col] == 'Q') return false
                if (row >= i && col >= i && board[row - i][col - i] == 'Q') return false
                if (row >= i && col + i < n && board[row - i][col + i] == 'Q') return false
            }
            return true
        }
    }
}
